File: typescript/src/streaming/WatchTogetherService.ts
/**
 * Watch-together synchronization service.
 *
 * The host drives playback (play, pause, seek, setSpeed) by broadcasting
 * WatchSync packets to all session members. Followers apply RTT compensation:
 *   adjustedPositionMs = position_ms + (Date.now() - sent_at_ms) * playback_speed
 *
 * Reactions are sent with WatchReaction (JSON) and delivered to all members.
 *
 * SPDX-License-Identifier: MIT
 */

import { DEFAULT_TTL } from "../constants.js";
import { MeshPacket } from "../protocol/MeshPacket.js";
import { PacketType } from "../protocol/PacketType.js";
import { IMeshSender } from "../routing/IMeshSender.js";

// ──────────────────────────────────────────────────────────────────
// Wire types (snake_case)
// ──────────────────────────────────────────────────────────────────

type WatchSyncKind = "play" | "pause" | "seek" | "speed" | "join" | "leave" | "end";

interface WatchSyncPayload {
  session_id: string;
  kind: WatchSyncKind;
  position_ms?: number;
  playback_speed?: number;
  sent_at_ms: number;
  content_id?: string;
}

interface WatchReactionPayload {
  session_id: string;
  reaction: string;
}

// ──────────────────────────────────────────────────────────────────
// Public types
// ──────────────────────────────────────────────────────────────────

export interface WatchSession {
  sessionId: string;
  hostUhid: string;
  contentId: string;
  members: Set<string>;
}

export interface SyncAppliedEvent {
  sessionId: string;
  kind: WatchSyncKind;
  positionMs: number;
  playbackSpeed: number;
  fromUhid: string;
}

export interface ReactionEvent {
  sessionId: string;
  fromUhid: string;
  reaction: string;
}

// ──────────────────────────────────────────────────────────────────
// Service
// ──────────────────────────────────────────────────────────────────

export class WatchTogetherService {
  private readonly sessions = new Map<string, WatchSession>();

  // Event callbacks
  onSessionInvited?: (session: WatchSession) => void;
  onSyncApplied?: (event: SyncAppliedEvent) => void;
  onReactionReceived?: (event: ReactionEvent) => void;
  onSessionEnded?: (sessionId: string) => void;

  constructor(private readonly sender: IMeshSender) {}

  // ──────────────── host session management ────────────────────────

  async inviteToSession(
    sessionId: string,
    contentId: string,
    memberUhids: string[],
  ): Promise<void> {
    if (!sessionId) throw new Error("sessionId must not be empty");
    const local = this.sender.localUhid;
    const members = new Set<string>([local, ...memberUhids]);

    const session: WatchSession = {
      sessionId,
      hostUhid: local,
      contentId,
      members,
    };
    this.sessions.set(sessionId, session);

    for (const uhid of memberUhids) {
      await this.sendSync(
        {
          session_id: sessionId,
          kind: "join",
          sent_at_ms: Date.now(),
          content_id: contentId,
        },
        uhid,
      );
    }
  }

  // ──────────────── host playback control ─────────────────────────

  async play(sessionId: string, positionMs: number): Promise<void> {
    await this.broadcastSync(sessionId, {
      session_id: sessionId,
      kind: "play",
      position_ms: positionMs,
      playback_speed: 1,
      sent_at_ms: Date.now(),
    });
  }

  async pause(sessionId: string, positionMs: number): Promise<void> {
    await this.broadcastSync(sessionId, {
      session_id: sessionId,
      kind: "pause",
      position_ms: positionMs,
      playback_speed: 0,
      sent_at_ms: Date.now(),
    });
  }

  async seek(sessionId: string, positionMs: number): Promise<void> {
    await this.broadcastSync(sessionId, {
      session_id: sessionId,
      kind: "seek",
      position_ms: positionMs,
      sent_at_ms: Date.now(),
    });
  }

  async setSpeed(sessionId: string, speed: number): Promise<void> {
    await this.broadcastSync(sessionId, {
      session_id: sessionId,
      kind: "speed",
      playback_speed: speed,
      sent_at_ms: Date.now(),
    });
  }

  // ──────────────── member actions ─────────────────────────────────

  async sendReaction(sessionId: string, reaction: string): Promise<void> {
    const session = this.sessions.get(sessionId);
    if (!session) return;

    const body = new TextEncoder().encode(
      JSON.stringify({ session_id: sessionId, reaction } satisfies WatchReactionPayload),
    );

    const local = this.sender.localUhid;
    for (const uhid of session.members) {
      if (uhid === local) continue;
      const packet = new MeshPacket();
      packet.type = PacketType.WatchReaction;
      packet.sourceUhid = local;
      packet.destinationUhid = uhid;
      packet.ttl = DEFAULT_TTL;
      packet.priority = 0;
      packet.payload = body;
      await this.sender.send(packet, uhid);
    }
  }

  // ──────────────── inbound packet routing ────────────────────────

  async onPacket(packet: MeshPacket): Promise<void> {
    switch (packet.type) {
      case PacketType.WatchSync:
        await this.handleSync(packet);
        break;
      case PacketType.WatchReaction:
        this.handleReaction(packet);
        break;
      default:
        break;
    }
  }

  // ──────────────── inspection ─────────────────────────────────────

  getSession(sessionId: string): WatchSession | undefined {
    return this.sessions.get(sessionId);
  }

  getActiveSessions(): WatchSession[] {
    return Array.from(this.sessions.values());
  }

  // ──────────────── private helpers ─────────────────────────────────

  private async handleSync(packet: MeshPacket): Promise<void> {
    let msg: WatchSyncPayload;
    try {
      msg = JSON.parse(new TextDecoder().decode(packet.payload)) as WatchSyncPayload;
    } catch {
      return;
    }
    if (!msg.session_id || !msg.kind) return;

    const { session_id: sessionId, kind } = msg;

    switch (kind) {
      case "join": {
        if (this.sessions.has(sessionId)) {
          // existing session — add sender as member
          this.sessions.get(sessionId)!.members.add(packet.sourceUhid);
          break;
        }
        const session: WatchSession = {
          sessionId,
          hostUhid: packet.sourceUhid,
          contentId: msg.content_id ?? "",
          members: new Set<string>([packet.sourceUhid, this.sender.localUhid]),
        };
        this.sessions.set(sessionId, session);
        this.onSessionInvited?.(session);
        break;
      }

      case "leave": {
        const session = this.sessions.get(sessionId);
        if (session) session.members.delete(packet.sourceUhid);
        break;
      }

      case "end": {
        this.sessions.delete(sessionId);
        this.onSessionEnded?.(sessionId);
        break;
      }

      case "play":
      case "pause":
      case "seek":
      case "speed": {
        const rawPositionMs = msg.position_ms ?? 0;
        const playbackSpeed = msg.playback_speed ?? 1;
        const sentAtMs = msg.sent_at_ms;

        // RTT compensation: advance position by transit time × speed
        const transitMs = Date.now() - sentAtMs;
        const compensatedPositionMs =
          kind === "pause"
            ? rawPositionMs
            : Math.max(0, rawPositionMs + transitMs * playbackSpeed);

        this.onSyncApplied?.({
          sessionId,
          kind,
          positionMs: compensatedPositionMs,
          playbackSpeed,
          fromUhid: packet.sourceUhid,
        });
        break;
      }

      default:
        break;
    }
  }

  private handleReaction(packet: MeshPacket): void {
    let msg: WatchReactionPayload;
    try {
      msg = JSON.parse(new TextDecoder().decode(packet.payload)) as WatchReactionPayload;
    } catch {
      return;
    }
    if (!msg.session_id || !msg.reaction) return;

    this.onReactionReceived?.({
      sessionId: msg.session_id,
      fromUhid: packet.sourceUhid,
      reaction: msg.reaction,
    });
  }

  private async broadcastSync(sessionId: string, payload: WatchSyncPayload): Promise<void> {
    const session = this.sessions.get(sessionId);
    if (!session) return;
    const local = this.sender.localUhid;
    for (const uhid of session.members) {
      if (uhid === local) continue;
      await this.sendSync(payload, uhid);
    }
  }

  private async sendSync(payload: WatchSyncPayload, toUhid: string): Promise<void> {
    const body = new TextEncoder().encode(JSON.stringify(payload));
    const packet = new MeshPacket();
    packet.type = PacketType.WatchSync;
    packet.sourceUhid = this.sender.localUhid;
    packet.destinationUhid = toUhid;
    packet.ttl = DEFAULT_TTL;
    packet.priority = 0;
    packet.payload = body;
    await this.sender.send(packet, toUhid);
  }
}

File: kotlin/src/main/kotlin/aethernet/streaming/WatchTogetherService.kt
// SPDX-License-Identifier: MIT

package aethernet.streaming

import aethernet.AetherNetConstants
import aethernet.protocol.MeshPacket
import aethernet.protocol.PacketType
import aethernet.routing.MeshSender
import aethernet.voice.JsonReader
import aethernet.voice.jsonEscape
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

// ─────────────────────────────────────────────────────────────────────────────
// Domain types
// ─────────────────────────────────────────────────────────────────────────────

enum class WatchTogetherKind { Play, Pause, Seek, Speed, Join, Leave, End }

data class WatchSyncEvent(
    val sessionId: UUID,
    val kind: WatchTogetherKind,
    val positionMs: Long?,
    val playbackSpeed: Double?,
    val sentAtMs: Long,
    val contentId: String?,
    val fromUhid: String
) {
    /**
     * RTT-compensated playback position.
     * position = positionMs + (now - sentAtMs) * playbackSpeed
     */
    fun compensatedPositionMs(speed: Double = playbackSpeed ?: 1.0): Long? {
        val pos = positionMs ?: return null
        val latencyMs = System.currentTimeMillis() - sentAtMs
        return pos + (latencyMs * speed).toLong()
    }
}

data class WatchReactionEvent(
    val sessionId: UUID,
    val reaction: String,
    val fromUhid: String
)

data class WatchSession(
    val id: UUID,
    val contentId: String,
    val members: MutableSet<String> = ConcurrentHashMap.newKeySet()
)

/**
 * Synchronized watch-together service.
 *
 * Sync packets use WatchSyncPayload JSON, reaction packets use WatchReactionPayload JSON.
 * RTT compensation is applied on receipt: compensated_position = positionMs + Δt * speed.
 */
class WatchTogetherService(private val sender: MeshSender) {

    private val sessions = ConcurrentHashMap<UUID, WatchSession>()

    /** Fired when a sync event (play, pause, seek, speed, join, leave, end) arrives. */
    var onSyncReceived: ((WatchSyncEvent) -> Unit)? = null

    /** Fired when a reaction arrives. */
    var onReactionReceived: ((WatchReactionEvent) -> Unit)? = null

    // ─────────────────────────────────────────────────────────────────────
    // Session management
    // ─────────────────────────────────────────────────────────────────────

    /**
     * Create or join a session locally (no network message). Call [inviteToSession]
     * to send invites to remote members.
     */
    fun createSession(sessionId: UUID, contentId: String): WatchSession {
        return sessions.getOrPut(sessionId) {
            WatchSession(id = sessionId, contentId = contentId).also {
                it.members.add(sender.localUhid)
            }
        }
    }

    /**
     * Send an invite+join signal to [memberUhids] for the given [sessionId].
     */
    suspend fun inviteToSession(sessionId: UUID, contentId: String, memberUhids: List<String>) {
        require(memberUhids.isNotEmpty()) { "memberUhids must not be empty" }
        val session = sessions.getOrPut(sessionId) {
            WatchSession(id = sessionId, contentId = contentId).also {
                it.members.add(sender.localUhid)
            }
        }
        for (uhid in memberUhids) {
            session.members.add(uhid)
            val payload = encodeWatchSync(
                sessionId = sessionId,
                kind = "join",
                positionMs = null,
                playbackSpeed = null,
                contentId = contentId
            )
            sender.send(watchSyncPacket(uhid, payload), uhid)
        }
    }

    // ─────────────────────────────────────────────────────────────────────
    // Playback control — broadcasts to all session members
    // ─────────────────────────────────────────────────────────────────────

    suspend fun play(sessionId: UUID, positionMs: Long) {
        broadcastSync(sessionId, "play", positionMs = positionMs)
    }

    suspend fun pause(sessionId: UUID, positionMs: Long) {
        broadcastSync(sessionId, "pause", positionMs = positionMs)
    }

    suspend fun seek(sessionId: UUID, positionMs: Long) {
        broadcastSync(sessionId, "seek", positionMs = positionMs)
    }

    suspend fun setSpeed(sessionId: UUID, playbackSpeed: Double) {
        broadcastSync(sessionId, "speed", playbackSpeed = playbackSpeed)
    }

    // ─────────────────────────────────────────────────────────────────────
    // Reactions
    // ─────────────────────────────────────────────────────────────────────

    suspend fun sendReaction(sessionId: UUID, reaction: String) {
        require(reaction.isNotEmpty()) { "reaction must not be empty" }
        val session = sessions[sessionId] ?: return
        val payload = encodeWatchReaction(sessionId, reaction)
        val packet = MeshPacket(
            type = PacketType.WatchReaction,
            sourceUhid = sender.localUhid,
            destinationUhid = "",
            ttl = AetherNetConstants.DEFAULT_TTL,
            priority = 16,
            payload = payload
        )
        for (member in session.members) {
            if (member != sender.localUhid) {
                sender.send(packet.copy(destinationUhid = member), member)
            }
        }
    }

    // ─────────────────────────────────────────────────────────────────────
    // Inbound packet dispatcher
    // ─────────────────────────────────────────────────────────────────────

    suspend fun handlePacket(packet: MeshPacket) {
        when (packet.type) {
            PacketType.WatchSync -> handleSync(packet)
            PacketType.WatchReaction -> handleReaction(packet)
            else -> {}
        }
    }

    // ─────────────────────────────────────────────────────────────────────
    // Private helpers
    // ─────────────────────────────────────────────────────────────────────

    private suspend fun broadcastSync(
        sessionId: UUID,
        kind: String,
        positionMs: Long? = null,
        playbackSpeed: Double? = null
    ) {
        val session = sessions[sessionId] ?: return
        val payload = encodeWatchSync(
            sessionId = sessionId,
            kind = kind,
            positionMs = positionMs,
            playbackSpeed = playbackSpeed
        )
        val packet = MeshPacket(
            type = PacketType.WatchSync,
            sourceUhid = sender.localUhid,
            destinationUhid = "",
            ttl = AetherNetConstants.DEFAULT_TTL,
            priority = 32,
            payload = payload
        )
        for (member in session.members) {
            if (member != sender.localUhid) {
                sender.send(packet.copy(destinationUhid = member), member)
            }
        }
    }

    private fun handleSync(packet: MeshPacket) {
        val json = String(packet.payload, Charsets.UTF_8)
        val sessionIdStr = JsonReader.readString(json, "session_id") ?: return
        val sessionId = runCatching { UUID.fromString(sessionIdStr) }.getOrNull() ?: return
        val kindStr = JsonReader.readString(json, "kind") ?: return
        val positionMs = JsonReader.readLong(json, "position_ms")
        val playbackSpeed = JsonReader.readDouble(json, "playback_speed")
        val sentAtMs = JsonReader.readLong(json, "sent_at_ms") ?: System.currentTimeMillis()
        val contentId = JsonReader.readString(json, "content_id")

        val kind = when (kindStr) {
            "play" -> WatchTogetherKind.Play
            "pause" -> WatchTogetherKind.Pause
            "seek" -> WatchTogetherKind.Seek
            "speed" -> WatchTogetherKind.Speed
            "join" -> WatchTogetherKind.Join
            "leave" -> WatchTogetherKind.Leave
            "end" -> WatchTogetherKind.End
            else -> return
        }

        // Maintain session membership on join/leave/end
        when (kind) {
            WatchTogetherKind.Join -> {
                val session = sessions.getOrPut(sessionId) {
                    WatchSession(id = sessionId, contentId = contentId ?: "")
                }
                session.members.add(packet.sourceUhid)
            }
            WatchTogetherKind.Leave -> {
                sessions[sessionId]?.members?.remove(packet.sourceUhid)
            }
            WatchTogetherKind.End -> {
                sessions.remove(sessionId)
            }
            else -> {}
        }

        val event = WatchSyncEvent(
            sessionId = sessionId,
            kind = kind,
            positionMs = positionMs,
            playbackSpeed = playbackSpeed,
            sentAtMs = sentAtMs,
            contentId = contentId,
            fromUhid = packet.sourceUhid
        )
        onSyncReceived?.invoke(event)
    }

    private fun handleReaction(packet: MeshPacket) {
        val json = String(packet.payload, Charsets.UTF_8)
        val sessionIdStr = JsonReader.readString(json, "session_id") ?: return
        val sessionId = runCatching { UUID.fromString(sessionIdStr) }.getOrNull() ?: return
        val reaction = JsonReader.readString(json, "reaction") ?: return

        onReactionReceived?.invoke(
            WatchReactionEvent(sessionId = sessionId, reaction = reaction, fromUhid = packet.sourceUhid)
        )
    }

    private fun watchSyncPacket(toUhid: String, payload: ByteArray) = MeshPacket(
        type = PacketType.WatchSync,
        sourceUhid = sender.localUhid,
        destinationUhid = toUhid,
        ttl = AetherNetConstants.DEFAULT_TTL,
        priority = 32,
        payload = payload
    )
}

// ─────────────────────────────────────────────────────────────────────────────
// Wire helpers
// ─────────────────────────────────────────────────────────────────────────────

private fun encodeWatchSync(
    sessionId: UUID,
    kind: String,
    positionMs: Long?,
    playbackSpeed: Double?,
    contentId: String? = null
): ByteArray {
    val sentAtMs = System.currentTimeMillis()
    val sb = StringBuilder()
    sb.append('{')
    sb.append("\"session_id\":\"").append(sessionId).append("\",")
    sb.append("\"kind\":\"").append(jsonEscape(kind)).append("\",")
    sb.append("\"sent_at_ms\":").append(sentAtMs)
    if (positionMs != null) sb.append(",\"position_ms\":").append(positionMs)
    if (playbackSpeed != null) sb.append(",\"playback_speed\":").append(playbackSpeed)
    if (contentId != null) sb.append(",\"content_id\":\"").append(jsonEscape(contentId)).append('"')
    sb.append('}')
    return sb.toString().toByteArray(Charsets.UTF_8)
}

private fun encodeWatchReaction(sessionId: UUID, reaction: String): ByteArray {
    val s = "{\"session_id\":\"$sessionId\",\"reaction\":\"${jsonEscape(reaction)}\"}"
    return s.toByteArray(Charsets.UTF_8)
}
